NDTensors/src/lib/GradedAxes/src/fusion.jl

```diff
@@ -53,21 +68,28 @@ function fuse_blocklengths(x::LabelledInteger, y::LabelledInteger)
   return labelled(unlabel(x) * unlabel(y), fuse_labels(label(x), label(y)))
 end
 
+flatten_maybe_nested(v::Vector{<:Integer}) = v
+flatten_maybe_nested(v::Vector{<:AbstractGradedUnitRange}) = reduce(vcat, blocklengths.(v))
+
 using BlockArrays: blockedrange, blocks
 function tensor_product(a1::AbstractBlockedUnitRange, a2::AbstractBlockedUnitRange)
-  blocklengths = map(vec(collect(Iterators.product(blocks(a1), blocks(a2))))) do x
-    return mapreduce(length, fuse_blocklengths, x)
-  end
+  maybe_nested = map(
+    it -> mapreduce(length, fuse_blocklengths, it),
+    Iterators.flatten((Iterators.product(blocks(a1), blocks(a2)),)),
+  )
+  blocklengths = flatten_maybe_nested(maybe_nested)
   return blockedrange(blocklengths)
 end
 
 function blocksortperm(a::AbstractBlockedUnitRange)
-  # TODO: Figure out how to deal with dual sectors.
-  # TODO: `rev=isdual(a)`  may not be correct for symmetries beyond `U(1)`.
-  ## return Block.(sortperm(nondual_sectors(a); rev=isdual(a)))
   return Block.(sortperm(blocklabels(a)))
 end
 
+# convention: sort UnitRangeDual according to nondual blocks
+function blocksortperm(a::UnitRangeDual)
+  return Block.(sortperm(blocklabels(nondual(a))))
+end
```

Comment:
This is kind of a matter of taste, but I would prefer deleting this version and instead make the version defined for `AbstractBlockedUnitRange` generic enough to handle both dual and non-dual cases, for example I think we can just change that version to be defined as:
```julia
  return Block.(sortperm(blocklabels(nondual(a))))
```
For unit ranges that aren't dual to begin with, `nondual` will be a no-op.

Reply:
I plan to deal with this in #1529

Reply:
Could you explain how you plan to deal with it in #1531?

Reply:
I plan to completely rewrite `UnitRandeDual` and replace most of its code with generic code for `AbstractGradedUnitRange`.

Reply:
My suggestion is to have one generic definition:
```julia
function blocksortperm(a::AbstractUnitRange)
  return Block.(sortperm(blocklabels(nondual(a))))
end
```
which would handle dual and non-dual ranges, that doesn't require introducing `AbstractGradedUnitRange`.

This might fail with non-graded unit ranges, but I think that is ok, or we can have a definition that for `BlockedUnitRanges`, `blocklabels(a)` outputs something like `fill(NoLabel(), blocklength(a))`, in which case sorting won't do anything and the permutation will be trivial.

Reply:
So my suggestion here, like my suggestion in https://github.com/ITensor/ITensors.jl/pull/1363#discussion_r1777757052, is to make a more generic definition for `AbstractUnitRange` (or maybe `AbstractBlockedUnitRange`), which doesn't explicitly require the existence of or make use of `AbstractGradedUnitRange`.

Reply:
Ok.